docs: add band matrix storage format documentation in the README of `lapack/base`

aayush0325 opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   adds band matrix storage format documentation in the README of `lapack/base` as discussed

## Related Issues

> Does this pull request have any related issues?

No. 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

aayush0325 committed:
docs: add banded matrices intro section in lapack/base

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/lapack/base/README.md:
```diff
@@ -22,6 +22,161 @@ limitations under the License.
+
+## Band Storage
+
+Many LAPACK routines work with banded matrices, which are stored compactly in two-dimensional arrays to save memory and improve computational efficiency. This section describes the different band storage formats used throughout the LAPACK library.
```
```suggestion
Many LAPACK routines work with banded matrices, which are stored compactly in two-dimensional arrays to save memory and improve computational efficiency. The following are the different band storage formats used throughout LAPACK.
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/lapack/base/README.md:
```diff
@@ -22,6 +22,161 @@ limitations under the License.
+
+## Band Storage
+
+Many LAPACK routines work with banded matrices, which are stored compactly in two-dimensional arrays to save memory and improve computational efficiency. The following are the different band storage formats used throughout LAPACK.
```
```suggestion
Many LAPACK routines work with banded matrices, which are stored compactly in two-dimensional arrays arranged in linear memory in order to save memory and improve computational efficiency. The following are the different band storage formats used throughout LAPACK.
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/lapack/base/README.md:
```diff
@@ -22,6 +22,161 @@ limitations under the License.
+**Upper Triangle Storage (UPLO = 'U'):**
+
+-   Element `A[i, j]` is stored in `AB[KD+1+i-j, j]`.
+-   Valid range for `i`: `max(1, j-KD) <= i <= j`
```
```suggestion
-   Valid range for `i`: `max(1, j-KD) <= i <= j`.
```

kgryte committed:
docs: add missing period

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/lapack/base/README.md:
```diff
@@ -22,6 +22,161 @@ limitations under the License.
+
+<!-- </equation> -->
+
+`AB` is a 3×5 matrix is 3×5 as `KD+1 = 2+1 = 3`. Similarly, given the following matrix `A`,
```
```suggestion
`AB` is a 3×5 matrix as `KD+1 = 2+1 = 3`. Similarly, given the following matrix `A`,
```

kgryte committed:
docs: fix copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
style: align ampersands

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/lapack/base/README.md:
```diff
@@ -22,6 +22,161 @@ limitations under the License.
+
+Triangular band matrices are stored in the same format as general band matrices:
+
+-   If `KL = 0`, the matrix is upper triangular.
```
It would be good to actually show matrices which are upper and lower triangular here.

kgryte commented on lib/node_modules/@stdlib/lapack/base/README.md:
Overall, this is looking good. It's a more readable version of https://netlib.org/lapack/lug/node124.html. That stated, one thing which needs to be changed is that currently everything is written using one-based indexing. We need to update everything to zero-based indexing.

kgryte replied:
This comment still needs to be resolved.

kgryte reviewed (changes requested):
One question I have: how do row- and column-major orders factor in here. For example, in [`dgbbrd`](https://github.com/OpenMathLib/OpenBLAS/blob/43fdff7f145736cb7fcccb81f3cabf9a5a84be76/lapack-netlib/LAPACKE/src/lapacke_dgbbrd.c#L35), the routine accepts a `layout` argument.

It may be good to also show how the band storage matrices are stored in linear memory for the two supported layouts. IIRC, this was a point of confusion during development.

kgryte commented on lib/node_modules/@stdlib/lapack/base/README.md:
````diff
@@ -22,6 +22,161 @@ limitations under the License.
+\right]
+```
+
+`AB` is a 3×5 matrix as `KD+1 = 2+1 = 3`. 
````
```suggestion
`AB` is a 3×5 matrix as `KD+1 = 2+1 = 3`.
```

kgryte committed:
style: remove trailing whitespace

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/lapack/base/README.md:
```diff
@@ -22,6 +22,161 @@ limitations under the License.
+
+Let `M = N = 5`, `KL = 2`, and `KU = 1`. Given an original band matrix `A`
+
+<!-- <equation class="equation" label="eq:band_matrix_a" align="center" raw="A = \left[\begin{array}{rrrrr}a_{11} & a_{12} & 0      & 0      & 0 \\a_{21} & a_{22} & a_{23} & 0      & 0 \\a_{31} & a_{32} & a_{33} & a_{34} & 0 \\0      & a_{42} & a_{43} & a_{44} & a_{45} \\0      & 0      & a_{53} & a_{54} & a_{55}\end{array}\right]" alt="Representation of band matrix A."> -->
```
In these comments, you do not need to preserve extra whitespace for aligning array elements. In fact, removing is preferred.

aayush0325 commented:
> One question I have: how do row- and column-major orders factor in here. For example, in [dgbbrd](https://github.com/OpenMathLib/OpenBLAS/blob/43fdff7f145736cb7fcccb81f3cabf9a5a84be76/lapack-netlib/LAPACKE/src/lapacke_dgbbrd.c#L35), the routine accepts a layout argument.

this is how i understand the row and column major interpretations of these matrices:

for column major:

```json
{
  "order": "column-major",
  "type": "banded",
  "M": 5,
  "N": 5,
  "KL": 2,
  "KU": 1,
  "A": [
    0.0, 0.0, 0.0, 1.1, 2.1, 3.1,
    0.0, 0.0, 1.2, 2.2, 3.2, 4.2,
    0.0, 0.0, 2.3, 3.3, 4.3, 5.3,
    0.0, 0.0, 3.4, 4.4, 5.4, 0.0,
    0.0, 0.0, 4.5, 5.5, 0.0, 0.0
  ],
  "A_mat": [
    [ 1.1, 1.2, 0.0, 0.0, 0.0 ],
    [ 2.1, 2.2, 2.3, 0.0, 0.0 ],
    [ 3.1, 3.2, 3.3, 3.4, 0.0 ],
    [ 0.0, 4.2, 4.3, 4.4, 4.5 ],
    [ 0.0, 0.0, 5.3, 5.4, 5.5 ]
  ],
}
```

for row major:

```json
{
  "order": "row-major",
  "type": "banded",
  "M": 5,
  "N": 5,
  "KL": 2,
  "KU": 1,
  "A": [
    0.0, 0.0, 0.0, 0.0, 0.0,
    0.0, 0.0, 0.0, 0.0, 0.0,
    0.0, 1.2, 2.3, 3.4, 4.5,
    1.1, 2.2, 3.3, 4.4, 5.5,
    2.1, 3.2, 4.3, 5.4, 0.0,
    3.1, 4.2, 5.3, 0.0, 0.0
  ],
  "A_mat": [
    [ 1.1, 1.2, 0.0, 0.0, 0.0 ],
    [ 2.1, 2.2, 2.3, 0.0, 0.0 ],
    [ 3.1, 3.2, 3.3, 3.4, 0.0 ],
    [ 0.0, 4.2, 4.3, 4.4, 4.5 ],
    [ 0.0, 0.0, 5.3, 5.4, 5.5 ]
  ]
}
```

<img width="200" height="168" alt="image" src="https://github.com/user-attachments/assets/256eec08-5759-4653-af87-a9fa2f3a0cec" />

 i am basically reading this column-by-column for col-major and row-by-row for row-major

aayush0325 committed:
chore: remove newlines

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented:
@aayush0325 What you are displaying is specific for LU correct with the extra cols/rows?

Regardless, I suggest making explicit with a couple of brief examples showing how banded matrices are stored according to layouts.

aayush0325 committed:
docs: add example

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/lapack/base/README.md:
```diff
@@ -30,13 +30,13 @@ Many LAPACK routines work with banded matrices, which are stored compactly in tw
 ### General Band Matrix Storage
 
-A general band matrix of size `M`-by-`N` with `KL` subdiagonals and `KU` superdiagonals is stored in a two-dimensional array `AB` with `KL+KU+1` rows and `N` columns.
+A general band matrix of size `M`-by-`N` with `KL` subdiagonals and `KU` superdiagonals is stored in a two-dimensional array `A` with `KL+KU+1` rows and `N` columns.
```
@aayush0325 What was the rationale for changing `A` to `AB` here and below? Now at, e.g., L101, you have

```
Element `A[i, j]` is stored in `A[1+i-j, j]`.
```

which no longer makes sense.

aayush0325 replied:
ran a silly search and replace here, sorry!! fixing

aayush0325 committed:
fix: fix search and replace error

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/lapack/base/README.md:
```diff
@@ -22,6 +22,213 @@ limitations under the License.
+A = \left[
+\begin{array}{rrrrr}
+    a_{11}   & a_{12}   & a_{13}   & 0        & 0 \\
+    {a_{12}} & a_{22}   & a_{23}   & a_{24}   & 0 \\
```
Why are some of these wrapped in `{}` (e.g., `a_{12}`)?

kgryte replied:
Applies here and elsewhere throughout this PR.

aayush0325 committed:
docs: fix brackets

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented:
A few comments still need to be resolved before we can finally get this one in. :)

aayush0325 committed:
docs: update indexing limits for 0 based indexing

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented:
cc @DivitJain26 Would you mind taking a look at this? Cheers!

kgryte committed:
Merge branch 'develop' into intro-lapack

DivitJain26 commented:
hello

MKL describes two methods of packing a band matrix according to the layout in [Matrix Storage Schemes for LAPACK Routines](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2023-0/matrix-storage-schemes-for-lapack-routines.html).

A more descriptive explanation can be found in the [cblas_?gbmv](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2024-1/cblas-gbmv.html#GUID-2FB74140-ADF9-4D82-8644-FD32025F6B71)

> Layout = CblasColMajor: Before entry, the leading (kl + ku + 1) by n part of the array a must contain the matrix of coefficients. This matrix must be supplied column-by-column, with the leading diagonal of the matrix in row (ku) of the array, the first super-diagonal starting at position 1 in row (ku - 1), the first sub-diagonal starting at position 0 in row (ku + 1), and so on. Elements in the array a that do not correspond to elements in the band matrix (such as the top left ku by ku triangle) are not referenced.

> Layout = CblasRowMajor: Before entry, the leading (kl + ku + 1) by m part of the array a must contain the matrix of coefficients. This matrix must be supplied row-by-row, with the leading diagonal of the matrix in column (kl) of the array, the first super-diagonal starting at position 0 in column (kl + 1), the first sub-diagonal starting at position 1 in row (kl - 1), and so on. Elements in the array a that do not correspond to elements in the band matrix (such as the top left kl by kl triangle) are not referenced.

This is true for BLAS routines as I have demonstrated here [Layout Dependent Band Matrix Packing in `cblas_?gbmv`](https://github.com/stdlib-js/stdlib/pull/6121#issuecomment-4390185998). 

However, it does not hold for LAPACK routines: routines that operate on band matrices expect the input matrix to be packed as defined for `Layout = CblasColMajor`, regardless of the chosen layout, as demonstrated [LAPACKE band routines always expect column-major style band packing; even when `layout = LAPACK_ROW_MAJOR`](https://github.com/stdlib-js/stdlib/pull/8083#issuecomment-4466540949).
- For `layout = LAPACK_COL_MAJOR`, this compact matrix should be supplied column-wise.
- For `layout = LAPACK_ROW_MAJOR`, the compact matrix should be supplied row-wise.

This contradicts [Matrix Storage Schemes for LAPACK Routines](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2023-0/matrix-storage-schemes-for-lapack-routines.html) and warrants further investigation.

The reason for this behaviour becomes clear when tracing the call chain. When `dgbbrd` is called, its wrapper [lapacke_dgbbrd.c](https://github.com/OpenMathLib/OpenBLAS/blob/43fdff7f145736cb7fcccb81f3cabf9a5a84be76/lapack-netlib/LAPACKE/src/lapacke_dgbbrd.c#L35) validates the inputs, and calls [lapacke_dgbbrd_work.c](https://github.com/OpenMathLib/OpenBLAS/blob/43fdff7f145736cb7fcccb81f3cabf9a5a84be76/lapack-netlib/LAPACKE/src/lapacke_dgbbrd_work.c) which in turn calls the native implementation [dgbbrd.c](https://github.com/OpenMathLib/OpenBLAS/blob/43fdff7f145736cb7fcccb81f3cabf9a5a84be76/lapack-netlib/SRC/dgbbrd.c). The native implementation has no layout parameter, as it operates exclusively in column-major style. Consequently, the wrapper simply transposes the input matrix when `layout = LAPACK_ROW_MAJOR`. This is why, even in the row-major case, the matrix must be packed according to the column-major convention and then supplied row-wise.

As mentioned in [Matrix Layout (Row Major and Column Major)](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2023-1/introduction-to-the-intel-oneapi-math-kernel.html#MATRIX-LAYOUT-ROW-MAJOR-AND-COLUMN-MAJOR).
>Matrix Layout (Row Major and Column Major)
The standard BLAS and LAPACK APIs require a Fortran layout for matrices (column major), where matrices are stored column-by-column in memory and the entries in each column are stored in consecutive memory locations. By default, oneMKL for DPC++ likewise assumes this matrix layout. The oneapi::mkl::blas::row_major namespace must be used for row major layout for BLAS. Row major layout is not supported directly for LAPACK, but you can transpose row major input matrices, call the desired DPC++ LAPACK routines, and then transpose the output matrices back to row major layout.

References:
[1] [Matrix Storage Schemes for LAPACK Routines](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2023-0/matrix-storage-schemes-for-lapack-routines.html).
[2] [cblas_?gbmv](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-c/2024-1/cblas-gbmv.html#GUID-2FB74140-ADF9-4D82-8644-FD32025F6B71)
[3] [Matrix Layout (Row Major and Column Major)](https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2023-1/introduction-to-the-intel-oneapi-math-kernel.html#MATRIX-LAYOUT-ROW-MAJOR-AND-COLUMN-MAJOR).
[4] [writing a banded matrix in a row major layout for lapack solver dgbsv](https://stackoverflow.com/questions/42750671/writing-a-banded-matrix-in-a-row-major-layout-for-lapack-solver-dgbsv)

DivitJain26 commented:
## LAPACKE band routines always expect column-major style band packing; even when `layout = LAPACK_ROW_MAJOR`

Demonstrates that LAPACKE band-matrix routines always expect the input packed using the column-major convention, regardless of the layout parameter.

```c
#include <stdio.h>
#include <lapacke.h>

int main() {

	lapack_int M = 4;
	lapack_int N = 4;

	lapack_int KL = 2;
	lapack_int KU = 1;

	/*
		A_mat: [
			[ 1.0,  2.0,  0.0,  0.0 ],
			[ 3.0,  4.0,  5.0,  0.0 ],
			[ 6.0,  7.0,  8.0,  9.0 ],
			[ 0.0, 10.0, 11.0, 12.0 ]
		]
	*/

	/*
		Compact band storage (column-major convention):
		A_compact: [
			[ 0.0,  2.0,  5.0,  9.0 ],
			[ 1.0,  4.0,  8.0, 12.0 ],
			[ 3.0,  7.0, 11.0,  0.0 ],
			[ 6.0, 10.0,  0.0,  0.0 ]
		]
	*/

	lapack_int LDAB = KL + KU + 1;

	double AB_col[4*4] = { 0.0, 1.0, 3.0, 6.0, 2.0, 4.0, 7.0, 10.0, 5.0, 8.0, 11.0, 0.0, 9.0, 12.0, 0.0, 0.0 };
	double AB_row[4*4] = { 0.0, 2.0, 5.0, 9.0, 1.0, 4.0, 8.0, 12.0, 3.0, 7.0, 11.0, 0.0, 6.0, 10.0, 0.0, 0.0 };

	double D_col[4];
	double E_col[4];
	double Q_col[4*4];
	double PT_col[4*4];
	double C_col[4*4];
	lapack_int info_col;
	double D_row[4];
	double E_row[4];
	double Q_row[4*4];
	double PT_row[4*4];
	double C_row[4*4];
	lapack_int info_row;

	info_col = LAPACKE_dgbbrd( LAPACK_COL_MAJOR, 'B', M, N, 0, KL, KU, AB_col, LDAB, D_col, E_col, Q_col, 4, PT_col, 4, C_col, 4 );
	info_row = LAPACKE_dgbbrd( LAPACK_ROW_MAJOR, 'B', M, N, 0, KL, KU, AB_row, LDAB, D_row, E_row, Q_row, 4, PT_row, 4, C_row, 4 );

	printf("\n=======Column Major========\n");
	printf("INFO = %d\n", info_col);

	printf("\nDiagonal D:\n");
	for(int i = 0; i < 4; i++) printf("%8.3f ", D_col[i]);

	printf("\n\nOff-diagonal E:\n");
	for(int i = 0; i < 3; i++) printf("%8.3f ", E_col[i]);

	printf("\n\nQ matrix:\n");
	for(int i = 0; i < 4; i++) {
		for(int j = 0; j < 4; j++) {
			printf("%8.3f ", Q_col[i + j*4]);
		}
		printf("\n");
	}

	printf("\nPT matrix:\n");
	for(int i = 0; i < 4; i++) {
		for(int j = 0; j < 4; j++) {
			printf("%8.3f ", PT_col[i + j*4]);
		}
		printf("\n");
	}

	printf("\n=======Row Major========\n");
	printf("INFO = %d\n", info_row);

	printf("\nDiagonal D:\n");
	for(int i = 0; i < 4; i++) printf("%8.3f ", D_row[i]);

	printf("\n\nOff-diagonal E:\n");
	for(int i = 0; i < 3; i++) printf("%8.3f ", E_row[i]);

	printf("\n\nQ matrix:\n");
	for(int i = 0; i < 4; i++) {
		for(int j = 0; j < 4; j++) {
			printf("%8.3f ", Q_row[i*4 + j]);
		}
		printf("\n");
	}

	printf("\nPT matrix:\n");
	for(int i = 0; i < 4; i++) {
		for(int j = 0; j < 4; j++) {
			printf("%8.3f ", PT_row[i*4 + j]);
		}
		printf("\n");
	}

	return 0;
}
```

Compile
```bash
gcc file_name.c -lmkl_rt -llapacke -lpthread -lm -ldl
```

Output
```bash

=======Column Major========
INFO = 0

Diagonal D:
   6.782   19.069    3.795    1.540 

Off-diagonal E:
  14.760    2.398   -0.123 

Q matrix:
   0.147   -0.055   -0.137    0.978 
   0.442   -0.060   -0.874   -0.193 
   0.885    0.039    0.460   -0.067 
   0.000    0.996   -0.078    0.045 

PT matrix:
   1.000    0.000    0.000    0.000 
   0.000    0.559    0.629    0.539 
   0.000   -0.327   -0.430    0.841 
   0.000    0.762   -0.647   -0.035 

=======Row Major========
INFO = 0

Diagonal D:
   6.782   19.069    3.795    1.540 

Off-diagonal E:
  14.760    2.398   -0.123 

Q matrix:
   0.147   -0.055   -0.137    0.978 
   0.442   -0.060   -0.874   -0.193 
   0.885    0.039    0.460   -0.067 
   0.000    0.996   -0.078    0.045 

PT matrix:
   1.000    0.000    0.000    0.000 
   0.000    0.559    0.629    0.539 
   0.000   -0.327   -0.430    0.841 
   0.000    0.762   -0.647   -0.035 
```
